R-package/R/xgboost.R

```diff
@@ -1,32 +1,1022 @@
+      )
+    )
+
+    if (NROW(y_names) == ncol(y)) {
+      out$metadata$y_names <- y_names
+    }
+
+  } else {
+    stop("Passed 'y' object with unsupported class: ", paste(class(y), collapse = ", "))
+  }
+
+  return(out)
+}
+
+process.row.weights <- function(w, lst_args) {
+  if (!is.null(w)) {
+    if ("label" %in% names(lst_args$dmatrix_args)) {
+      nrow_y <- NROW(lst_args$dmatrix_args$label)
+    } else if ("label_lower_bound" %in% names(lst_args$dmatrix_args)) {
+      nrow_y <- length(lst_args$dmatrix_args$label_lower_bound)
+    } else {
+      stop("Internal error.")
+    }
+    if (!is.numeric(w)) {
+      w <- as.numeric(w)
+    }
+    if (length(w) != nrow_y) {
+      stop(
+        "'weights' must be a 1-d vector with the same length as 'y' (",
+        length(w), " vs. ", nrow_y, ")."
+      )
+    }
+    lst_args$dmatrix_args$weight <- w
+  }
+  return(lst_args)
+}
+
+check.nthreads <- function(nthreads) {
+  if (is.null(nthreads)) {
+    return(1L)
+  }
+  if (!inherits(nthreads, c("numeric", "integer")) || !NROW(nthreads)) {
+    stop("'nthreads' must be a positive scalar value.")
+  }
+  if (length(nthreads) > 1L) {
+    nthreads <- head(nthreads, 1L)
+  }
+  if (is.na(nthreads) || nthreads < 1) {
+    stop("Passed invalid 'nthreads': ", nthreads)
+  }
+  if (is.numeric(nthreads)) {
+    if (floor(nthreads) != nthreads) {
+      stop("'nthreads' must be an integer.")
+    }
+  }
+  return(as.integer(nthreads))
+}
+
+check.can.use.qdm <- function(x, params) {
+  if (inherits(x, "sparseMatrix") && !inherits(x, "dgRMatrix")) {
+    return(FALSE)
+  }
+  if ("booster" %in% names(params)) {
+    if (params$booster == "gblinear") {
+      return(FALSE)
+    }
+  }
+  if ("tree_method" %in% names(params)) {
+    if (params$tree_method %in% c("exact", "approx")) {
+      return(FALSE)
+    }
+  }
+  return(TRUE)
+}
+
+process.x.and.col.args <- function(
+  x,
+  monotone_constraints,
+  interaction_constraints,
+  feature_weights,
+  lst_args
+) {
+  if (is.null(x)) {
+    stop("'x' cannot be NULL.")
+  }
+  if (inherits(x, "xgb.DMatrix")) {
+    stop("Cannot pass 'xgb.DMatrix' as 'x' to 'xgboost()'. Try 'xgb.train()' instead.")
+  }
+  supported_x_types <- c("data.frame", "matrix", "dgCMatrix", "dgRMatrix")
+  if (!inherits(x, supported_x_types)) {
+    stop(
+      "'x' must be one of the following classes: ",
+      paste(supported_x_types, collapse = ", "),
+      ". Got: ",
+      paste(class(x), collapse = ", ")
+    )
+  }
+
+  if (NROW(feature_weights)) {
+    if (is.list(feature_weights)) {
+      feature_weights <- unlist(feature_weights)
+    }
+    if (!inherits(feature_weights, c("numeric", "integer"))) {
+      stop("'feature_weights' must be a numeric vector or named list matching to columns of 'x'.")
+    }
+    if (length(feature_weights) != ncol(x)) {
+      stop(
+        "'feature_weights' does not match in length with columns of 'x' (",
+        length(feature_weights), " vs. ", ncol(x), ")."
```

Comment:
Is it possible that we can use the check in c++ https://github.com/dmlc/xgboost/blob/513d7a7d84678fc4e75c18899fe52263258f0290/src/data/data.cc#L785 ?